### ISSUE TITLE
remove schedule, add trip

### ISSUE THREAD

**Issue by juliuste** [public-transport/friendly-public-transport-format#42]

**Comment by juliuste** [comment]
⬆️ Fixed the issue text now.

**Comment by derhuerst** [comment]
> Schedule information is - sadly - really rare, at least for public transport in Europe, in my experience.

I disagree mostly. I'd say that, if anywhere, *in Europe* scheduled public transportation seems to be common. From my subjective experience, other parts often rely more on demand-responsive and/or less centrally organized means of transportation.

> Because of how `schedule` interacts with `route` (and `line`) types in the current spec, using these other types without `schedule` becomes unnecessarily complicated in a vast majority of use cases.

- I can see that computing `schedule`s from e.g. GTFS can be cumbersome. But this structure is what one often computes *implicitly* trying answer questions like "how often does this bus run in the next hours?". So why not have/keep a format for it?
- A reasonable number of GTFS feeds don't just provide the raw `trip` + `line` info, but also schedule data. As GTFS is the de-facto standard for static transportation data, we may want to have a related (but useful!) structure in FPTF.
- Some modern routing algorithms (not CSA) make use of information like this.

> - It's unclear where to indicate *scheduled* and *realtime* time information, e.g. in `journey`s. It would make sense to always keep *scheduled* information in `schedule`s, but that's often not possible due to a lack of data.

Good point, they are kind of redundant, due to the practical limitation of APIs, where you often only get a very sparse view on the data. Let's see what we can do about it.

> - I am not sure if it even makes sense to say that multiple trips at different times share one schedule. What makes two schedules the same? Is it really just that vehicles stop at the same places and take roughly the same time? Even very consistent metro lines often have multiple schedules since the trains take longer depending on how much people want to get in and out of the trains. I feel like the main reason for us to still use `schedule`s is not that they inherently make sense, but rather that in public transport our smallest unit of time is still `minute`s so that it just seems like some trips share one common schedule even though they actually don't. `schedule`s seem like some kind of *hack* everyone has just gotten used to. I have to admit that this is a very theoretical point of criticism, but we might at least want to consider it if there is another alternative.

Our intial intention – and what I think we should focus on, given that we created yet another format, and call it *friendly* – was to **accept the current reality** (e.g. sparse views on the data, scheduled but delayed systems), **but nudge users towards a more flexible and consistent future**. To give some examples:

- a `type` field on every item – not strictly necessary, but makes discovery and debugging of APIs easier
- `mode`s & `product`s – we allow coarse *and* fine-grained info
- `regions` – we discourage `HAMBURG`/`Berlin (alle Stationen)` hacks
- ISO 8601 strings instead of timestamps – we accept data without timezones, but allow providing them

To be more specific, I definitely agree with your theoretical point about `schedule`s and that they're often a hack. We have to decide though **where we want to position FPTF on the practicality/consistency spectrum**.

In practice, you can find `schedule`s at almost every bus stop and train station in Europe. In practice, people expect vehicles to take a more or less consistent time to travel from A to B. In practice, routing engines depend on `schedule`-like information.

> Another caveat for *FPTF* is that the lack of a `trip` type makes it harder to integrate our format into other systems since trips are mostly - e.g. in *GTFS* or *linked connections* - seen as a basic block of transport systems, alongside `station`s and `line`s. This is also highly relevant for unscheduled services that might be growing in the following years.

> […] at the moment *FPTF* is almost exclusively used for non-static applications, and even for static data dumps it might actually bring some benefits. For example: Routing algorithms like *CSA* operate on sorted lists of trips as the most basic type of transport relations.

I agree, we definitely need a `trip`. This concept exists explicitly in GTFS and HAFAS, and implicitly in many routing algorithms and most transportation APIs/apps I've seen.

> When we were originally “designing” the format, we wanted to find some tradeoff between usability and size. One could argue that introducing trips drastically increases the space needed to store e.g. a dump of given system.

I think storage space is not a problem. What I would rather use as a metric – next to usability obviously – is "raw-ness" or readyness to consume it in a meaningful way. People who want to optimize their FPTF data sets will be able to encode it more efficiently.

Given that I argue to keep `schedule`, people who want more compact – and less raw – data sets would use `schedule` objects. People who want less computational effort and/or do CSA-routing would use `trip` objects.

---

> Long story short, I propose the following changes for *FPTF 2*, please express your opinions!
>
> 1. Remove the `schedule` type.
> 2. Introduce a `trip` type […]
> 3. Change the `journey` type as follows:
> 	- Remove `leg.stopovers`
> 	- Remove `leg.schedule`
> 	- Add `leg.trip`
> 4. Implement #33 (scheduled/realtime when attributes) for `stopover`s and `journey`s

I will sum up my response as well:

- **I would like to keep `schedule` or something comparable**, for the mentioned reasons.
- **Agree that we need `trip`**, and if we add it, we should **adapt `schedule` in a clever way so that it doesn't overlap with `trip` that much**.
- Limiting the format by removing `leg.stopovers`, for the sake of theoretical consistency, doesn't seem worth it in this case. We need to *keep* a way of specifying sparse data (e.g. "i know stopovers from A to B, but not where the bus has been before or will be after" my journey). – Except if we find a clever way to specify a *sparse* `trip`.

Let's design an solution, appropriate to our "accept the reality, but nudge towards clean, consistent, meaningful markup" motive.

**Comment by juliuste** [comment]
> I disagree mostly. I'd say that, if anywhere, in Europe scheduled public transportation seems to be common. From my subjective experience, other parts often rely more on demand-responsive and/or less centrally organized means of transportation.

I meant that the schedule information is mostly not (digitally) available, not that it doesn't exist at all.

> I agree, we definitely need a `trip`. This concept exists explicitly in GTFS and HAFAS, and implicitly in many routing algorithms and most transportation APIs/apps I've seen.

👍 

> - I can see that computing `schedule`s from e.g. GTFS can be cumbersome. But this structure is what one often computes *implicitly* trying answer questions like "how often does this bus run in the next hours?". So why not have/keep a format for it?

But from my understanding `schedule`s don't actually provide this information… They just say which services that take the same time between stations run in the next few hours. If there is another trip of the same line that happens to wait for just one more minute at some stop, you would probably want that trip to be included in the answer to your question "how often does this bus run in the next hours?" as well, right?

I get your point, but I see the problem here: The schedules we see at bus stops or questions like the one you just mentioned always describe a schedule **at a specific station** while our current FPTF approach to schedules is station-independent.

> - A reasonable number of GTFS feeds don't just provide the raw `trip` + `line` info, but also schedule data. As GTFS is the de-facto standard for static transportation data, we may want to have a related (but useful!) structure in FPTF.

Do you refer to GTFS `services` here? If so, aren't they quite different from our `schedules`? (Since they describe services in relation to the day (of week) which is more consistent than our schedules which describe specific travel times.)

> - Limiting the format by removing `leg.stopovers`, for the sake of theoretical consistency, doesn't seem worth it in this case. We need to *keep* a way of specifying sparse data (e.g. "i know stopovers from A to B, but not where the bus has been before or will be after" my journey). – Except if we find a clever way to specify a *sparse* `trip`.

👍 

> - **I would like to keep `schedule` or something comparable**, for the mentioned reasons.

Overall, I get your point, however I feel like schedules in their current state (`1.0.2`) are not viable and we should change them for the given reasons (like scheduled/realtime information redundancy, trip/schedule redundancy etc.)
